File: songa_mobility_phase_2/songa_app_integration/report/commission_ledger_status/commission_ledger_status.py
# Copyright (c) 2026, Lucky Tsuma and contributors
# For license information, please see license.txt

import frappe


def execute(filters=None):
	filters = frappe._dict(filters or {})
	data = get_data(filters)
	return get_columns(), data, None, get_chart(data)


def get_columns():
	return [
		{"fieldname": "posting_date", "label": "Date", "fieldtype": "Date", "width": 100},
		{
			"fieldname": "name",
			"label": "Ledger Entry",
			"fieldtype": "Link",
			"options": "Driver Commission Ledger",
			"width": 170,
		},
		{"fieldname": "driver", "label": "Driver", "fieldtype": "Link", "options": "Driver", "width": 120},
		{"fieldname": "driver_name", "label": "Driver Name", "fieldtype": "Data", "width": 150},
		{"fieldname": "transaction_type", "label": "Type", "fieldtype": "Data", "width": 100},
		{"fieldname": "usage", "label": "Usage", "fieldtype": "Data", "width": 150},
		{"fieldname": "amount", "label": "Amount", "fieldtype": "Currency", "width": 120},
		{"fieldname": "workflow_state", "label": "Status", "fieldtype": "Data", "width": 110},
		{
			"fieldname": "journal_entry",
			"label": "Journal Entry",
			"fieldtype": "Link",
			"options": "Journal Entry",
			"width": 160,
		},
	]


def get_data(filters):
	conditions = {}
	if filters.get("company"):
		conditions["company"] = filters.get("company")
	if filters.get("driver"):
		conditions["driver"] = filters.get("driver")
	if filters.get("transaction_type"):
		conditions["transaction_type"] = filters.get("transaction_type")
	if filters.get("workflow_state"):
		conditions["workflow_state"] = filters.get("workflow_state")
	else:
		# Default to the live (non-cancelled) approval pipeline.
		conditions["docstatus"] = ["<", 2]

	if filters.get("from_date") and filters.get("to_date"):
		conditions["posting_date"] = ["between", [filters.get("from_date"), filters.get("to_date")]]
	elif filters.get("from_date"):
		conditions["posting_date"] = [">=", filters.get("from_date")]
	elif filters.get("to_date"):
		conditions["posting_date"] = ["<=", filters.get("to_date")]

	return frappe.get_all(
		"Driver Commission Ledger",
		filters=conditions,
		fields=[
			"name",
			"posting_date",
			"driver",
			"driver_name",
			"transaction_type",
			"usage",
			"amount",
			"workflow_state",
			"journal_entry",
		],
		order_by="posting_date desc, creation desc",
	)


def get_chart(data):
	counts = {}
	for row in data:
		state = row.get("workflow_state") or "Unknown"
		counts[state] = counts.get(state, 0) + 1
	if not counts:
		return None

	labels = list(counts.keys())
	return {
		"data": {"labels": labels, "datasets": [{"name": "Entries", "values": [counts[label] for label in labels]}]},
		"type": "donut",
	}

File: songa_mobility_phase_2/songa_app_integration/report/commission_ledger_status/commission_ledger_status.js
// Copyright (c) 2026, Lucky Tsuma and contributors
// For license information, please see license.txt

frappe.query_reports["Commission Ledger Status"] = {
	filters: [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("Company"),
		},
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			default: frappe.datetime.add_months(frappe.datetime.now_date(), -3),
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			default: frappe.datetime.now_date(),
		},
		{
			fieldname: "workflow_state",
			label: __("Status"),
			fieldtype: "Select",
			options: ["", "Pending", "Approved", "Rejected", "Cancelled"].join("\n"),
		},
		{
			fieldname: "transaction_type",
			label: __("Transaction Type"),
			fieldtype: "Select",
			options: ["", "Allocation", "Deduction"].join("\n"),
		},
		{
			fieldname: "driver",
			label: __("Driver"),
			fieldtype: "Link",
			options: "Driver",
		},
	],
	formatter: function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		if (data && column.fieldname === "workflow_state") {
			const colorMap = {
				Approved: "green",
				Rejected: "red",
				Cancelled: "gray",
				Pending: "orange",
			};
			const color = colorMap[value];
			if (color) {
				value = `<span style="color: ${color}; font-weight: 500;">${value}</span>`;
			}
		}

		return value;
	},
};
